src/internal/pfsdb/commit_provenance.go

```diff
@@ -0,0 +1,181 @@
+package pfsdb
+
+import (
+	"context"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+	pfsserver "github.com/pachyderm/pachyderm/v2/src/server/pfs"
+)
+
+// returns the commit of a certain repo in a commit set.
+func ResolveCommitProvenance(tx *pachsql.Tx, repo *pfs.Repo, commitSet string) (*pfs.Commit, error) {
+	cs, err := CommitSetProvenance(tx, commitSet)
+	if err != nil {
+		return nil, err
+	}
+	for _, c := range cs {
+		if RepoKey(c.Repo) == RepoKey(repo) {
+			return c, nil
+		}
+	}
+	return nil, pfsserver.ErrCommitNotFound{Commit: &pfs.Commit{Repo: repo, ID: commitSet}}
+}
+
+// CommitSetProvenance returns all the commit IDs that are in the provenance
+// of all the commits in this commit set.
+func CommitSetProvenance(tx *pachsql.Tx, id string) ([]*pfs.Commit, error) {
+	q := `
+          WITH RECURSIVE prov(from_id, to_id) AS (
+            SELECT from_id, to_id 
+            FROM pfs.commit_provenance JOIN pfs.commits ON int_id = from_id 
+            WHERE commit_set_id = $1
+           UNION ALL
+            SELECT cp.from_id, cp.to_id
+            FROM prov p, pfs.commit_provenance cp
+            WHERE cp.from_id = p.to_id
+          )
+          SELECT DISTINCT commit_id
```

Comment:
It would probably be better to use the `int_id` here. Same for the subvenance query.